dockstore-webservice/src/main/java/io/dockstore/webservice/resources/UserResource.java

```diff
@@ -1022,18 +1070,47 @@ public List<Workflow> addUserToDockstoreWorkflows(@ApiParam(hidden = true) @Para
                 .collect(Collectors.toList());
 
         scTokens.forEach(token -> {
-            SourceCodeRepoInterface sourceCodeRepo =  SourceCodeRepoFactory.createSourceCodeRepo(token);
-            Map<String, String> gitUrlToRepositoryId = sourceCodeRepo.getWorkflowGitUrl2RepositoryId();
-            Set<String> organizations = gitUrlToRepositoryId.values().stream().map(repository -> repository.split("/")[0]).collect(Collectors.toSet());
-
-            organizations.forEach(organization -> {
-                List<Workflow> workflowsWithoutuser = workflowDAO.findByOrganizationWithoutUser(token.getTokenSource().getSourceControl(), organization, user);
-                workflowsWithoutuser.forEach(workflow -> workflow.addUser(user));
-            });
+            final SourceCodeRepoInterface sourceCodeRepo =  SourceCodeRepoFactory.createSourceCodeRepo(token);
+            final SourceControl sourceControl = token.getTokenSource().getSourceControl();
+            addUserToReposInOrgsWhereUserIsAMember(user, sourceControl, sourceCodeRepo);
+            addUserToReposWithRepoLevelMembership(user, sourceControl, sourceCodeRepo);
         });
         return convertMyWorkflowsToWorkflow(this.bioWorkflowDAO.findUserBioWorkflows(user.getId()));
     }
 
+    /**
+     * Adds the user to all workflows whose repos the user direct access to. For example, if a user has been
+     * granted GitHub permissions to the GitHub organization "myorg" and the repository "anotherOrg/repo",
+     * this method will add the user to "anotherOrg/repo", only (if not already present).
+     * @param user
+     * @param sourceControl
+     * @param sourceCodeRepo
+     */
+    private void addUserToReposWithRepoLevelMembership(final User user, final SourceControl sourceControl,
+        final SourceCodeRepoInterface sourceCodeRepo) {
+        sourceCodeRepo.getRepoLevelAccessRepositories().forEach(gitRepo ->
```

Comment:
Technically, I don't think this is streaming, it's a short-hand for iteration,  as is your `for`. 

I get your broader point about one-liners, and will comment in that PR, but here, to me, this reads better: if it's a for loop, then I have to look inside it to see if there are breaks, conditions, etc, whereas with the `forEach` I just focus on what it's doing. The 2 `forEach`s also save the two time use of the word `final`, which I believe you think I overuse, but which I'm (overly?) zealous about, as every time I read code with a mutable variable declared, my mind goes into "where/when does it get modified?" mode.

But we're getting into personal preferences, and it could be worth on having a team coding style that we all follow.

Reply:
> Technically, I don't think this is streaming, it's a short-hand for iteration, as is your `for`.
> 
> I get your broader point about one-liners, and will comment in that PR, but here, to me, this reads better: if it's a for loop, then I have to look inside it to see if there are breaks, conditions, etc, whereas with the `forEach` I just focus on what it's doing. The 2 `forEach`s also save the two time use of the word `final`, which I believe you think I overuse, but which I'm (overly?) zealous about, as every time I read code with a mutable variable declared, my mind goes into "where/when does it get modified?" mode.
> 
> But we're getting into personal preferences, and it could be worth on having a team coding style that we all follow.

Yeah, my characterizing it as streaming was wrong, I think I ported the comment from somewhere else I was gonna use it and didn't change it. :)

Fwiw, it took me a significantly longer time (5-10x?) to understand the `forEach` based code than it would have if it was just plain properly-indented `for` statements.  I suspect a lot of it has to do with the indentation and the particular syntax involved (mix of parens and curly braces versus all parens).

Reply:
In the spirit of good-natured debate, I must [regretably] note that lambda arguments are by default non-`final` and can be assigned in the lambda body just like the loop variable in a `for`.  The following code compiles and prints "moo":
```
List<String> list = List.of("abc");
list.forEach(l -> System.out.println(l = "moo"));
```

Reply:
Touché -- I'd just assumed and never thought to try it.

Reply:
It should work your way. :)
In Steve's Programming Language (S), variables are `final` by default, with the option to make them mutable by appending an exclamation point to their name.